src/components/Pagination/hooks/usePagination.ts

```diff
@@ -0,0 +1,59 @@
+import React, { useMemo } from "react";
+import { BaseUsePaginationProps } from "../type";
+
+export const DOTS = "...";
+
+// WA: Create Array with certain range based on data length
+export const range = (start: number, end: number) => {
+  let length = end - start + 1;
+  return Array.from({ length }, (_, idx) => idx + start);
+};
+
+export const usePagination = (props: BaseUsePaginationProps) => {
+  const { totalPage, pageSize, siblingCount = 1, currentPage } = props;
+  const paginationRange = useMemo(() => {
+    // WA: Pages count is determined as siblingCount + firstPage + lastPage + currentPage + 2*DOTS
+    const totalPageNumbers = siblingCount + 5;
+    /*  
+      WA: If the number of pages is less than the page numbers we want to show in our
+      paginationComponent, we return the range [1..totalPage]
+    */
+    if (totalPageNumbers >= totalPage) {
+      return range(1, totalPage);
+    }
+
+    const leftSiblingIndex = Math.max(currentPage - siblingCount, 1);
+    const rightSiblingIndex = Math.min(currentPage + siblingCount, totalPage);
+
+    /*
+      WA: We do not want to show dots if there is only one position left 
+      after/before the left/right page count as that would lead to a change if our Pagination
+      component size which we do not want
+    */
+    const shouldShowLeftDots = leftSiblingIndex > 2;
+    const shouldShowRightDots = rightSiblingIndex < totalPage - 2;
+
+    const firstPageIndex = 1;
+    const lastPageIndex = totalPage;
+
+    if (!shouldShowLeftDots && shouldShowRightDots) {
+      let leftItemCount = 2 + 2 * siblingCount;
+      let leftRange = range(1, leftItemCount);
+
+      return [...leftRange, DOTS, totalPage];
+    }
+
+    if (shouldShowLeftDots && !shouldShowRightDots) {
+      let rightItemCount = 2 + 2 * siblingCount;
+      let rightRange = range(totalPage - rightItemCount + 1, totalPage);
+      return [firstPageIndex, DOTS, ...rightRange];
+    }
+
+    if (shouldShowLeftDots && shouldShowRightDots) {
+      let middleRange = range(leftSiblingIndex, rightSiblingIndex);
+      return [firstPageIndex, DOTS, ...middleRange, DOTS, lastPageIndex];
+    }
+  }, [totalPage, pageSize, siblingCount, currentPage]);
+
+  return paginationRange;
```

Comment:
hmmm so to sum up ada beberapa case:
- case kalo show semua (number of pages exceed number of page numbers)
- case kalo dots di kiri, tapi ga di kanan (not sure how this works yet, karena blm ngerti siblingCount)
- case kalo dots di kanan, tapi ga di kiri (not sure how this works yet, karena blm ngerti siblingCount)
- case kalo dots di kiri dan kanan (not sure how this works yet, karena blm ngerti siblingCount)

begitukah?

Reply:
yep, siblingCount udah dijelasin diatas ya

Reply:
okeh, perfect